virtual_ecosystem/models/abiotic/wind.py

```diff
@@ -302,27 +303,76 @@ def mix_and_ventilate(
     """
 
     input_variable_mixed = input_variable.copy()
-    for i in range(2, len(input_variable) - 1):
-        flux_up = (
-            mixing_coefficient[i - 1]
-            * (input_variable[i - 1] - input_variable[i])
-            / layer_thickness[i]
-        )
-        flux_down = (
-            mixing_coefficient[i + 1]
-            * (input_variable[i + 1] - input_variable[i])
-            / layer_thickness[i]
-        )
-        change_input_variable = (
-            (flux_up + flux_down) * time_interval / layer_thickness[i]
-        )
-        input_variable_mixed[i] += change_input_variable
-
-    # Ventilation at top
-    input_variable_mixed[0] += (
-        ventilation_rate * (input_variable[0] - input_variable[1]) * time_interval
-    )
-
+    n_layers, n_cells = input_variable.shape
+
+    for j in range(n_cells):
+        for i in range(1, n_layers - 1):  # skip top and bottom
+            center_val = input_variable[i, j]
+
+            if not np.isfinite(center_val):
+                continue
+
+            # ---- Upward flux ----
+            upper_i = i - 1
+            while upper_i >= 0 and not np.isfinite(input_variable[upper_i, j]):
+                upper_i -= 1
+
+            if upper_i >= 0 and np.isfinite(mixing_coefficient[upper_i, j]):
```

Comment:
Should this be indented within the while block above? Because the while block above doesn't seem to do anything other than decrease `upper_i`

Reply:
Good question, but no, the block should not be indented. The purpose of the `while` loop is only to find the nearest valid layer above/below the current one. Once it finds a finite value (or exits when it can't), the code that follows uses that upper_i value to calculate the flux.

From what I understand, if I indented the `if` block inside the `while`, it would try to calculate the flux on every invalid layer on the way up which is what we want to avoid.not what we want. You only want to calculate it once you've found a valid upper_i, or skip it if none is found.

I added comments inline and also added an explanation in the function docstring, I hope that makes sense

Reply:
Ahh right, I think it might be better to find these indices using `numpy` if possible. Generally loops in python aren't very efficient, so doing stuff in vectorised form is better (though I don't think it's a massive deal in this case).

I *think* this loop can be replaced with

```upper_i  = np.argmax(np.isfinite(input_variable[:, j]))```

which finds the first finite index in the array `input_variable[:, j]`. The lower two loops seem a bit more tricky

Reply:
I think the looping over cells can be replaced with array-based calculation - and I think that should be much faster, which will be critical here because it is having to loop at such a short time scale. I'm seeing if I can figure it out but I've added another couple of questions.